### PR TITLE
added fix to synchronise vpnservice and ipsec site connection requests.

### PR DESCRIPTION
**Pull request by sdharav**
@njagadish @rajendramachani  please review.

### CODE REVIEW

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py [hunk 1]
```diff
@@ -183,14 +193,46 @@ def create_ipsec_site_connection(self, context, ipsec_site_connection):
...
+                    resource=ipsec_site_connection,
+                    reason='create', service_vendor=service_vendor)
+                break
+            elif vpnservice['status'] == ERROR:
+                LOG.error('updating ipsec_site_conn to ERROR state')
+                self._update_ipsec_conn_state(context, ipsec_site_connection)
+                break
+            time.sleep(5)
+            endtime = time.time()
+        else:
```
Comment (njagadish) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py:212]: Couldn't find a matching if case for this else block

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py [hunk 2]
```diff
@@ -183,14 +193,46 @@ def create_ipsec_site_connection(self, context, ipsec_site_connection):
...
+            LOG.error('updating ipsec_site_conn to ERROR state')
+            self._update_ipsec_conn_state(context, ipsec_site_connection)
```
Comment (njagadish) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py:214]: Pass the state also as argument.

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py [hunk 3]
```diff
@@ -216,3 +258,15 @@ def create_vpnservice(self, context, vpnservice):
             rsrc_id=vpnservice['id'],
             resource=vpnservice,
             reason='create', service_vendor=service_vendor)
+
+    def delete_vpnservice(self, context, vpnservice):
```
Comment (njagadish) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py:256]: Why is VPN service delete passed to the configurtor driver? Are we handling this?

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py [hunk 4]
```diff
@@ -70,6 +75,11 @@ def ipsec_site_conn_deleted(self, context, ipsec_site_conn_id):
         plugin = self.driver.service_plugin
         plugin._delete_ipsec_site_connection(context, ipsec_site_conn_id)
 
+    def vpnservice_deleted(self, context, **kwargs):
```
Comment (njagadish) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py]: Don't see vpnservice_deleted callback in VPN agent code of configurator. Why is this added?

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py [hunk 5]
```diff
@@ -29,6 +31,9 @@
 
 BASE_VPN_VERSION = '1.0'
 AGENT_TYPE_VPN = 'NFP Vpn agent'
+ACTIVE = 'ACTIVE'
+ERROR = 'ERROR'
+TIMEOUT = 45
```
Comment (njagadish) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py]: Waiting for 45 seconds between VPN service create and IPSec connection create is too long as configurator is not processing anything. Let us keep this to 30 sec.

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py (hunk 2 above)
Comment (sdharav) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py:214]: that we are doing in the called method

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py (hunk 3 above)
Comment (sdharav) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py:256]: That for non GBP communications we get one delete vpn service request

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py (hunk 1 above)
Comment (sdharav) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py:212]: while loop

File: gbpservice/nfp/configurator/drivers/vpn/vyos/vyos_vpn_driver.py [hunk 6]
```diff
@@ -745,7 +745,8 @@ def __init__(self, conf):
         self.conf = conf
         self.handlers = {
             'vpn_service': {
-                'create': self.create_vpn_service},
+                'create': self.create_vpn_service,
+                'delete': self.delete_vpn_service},
```
Comment (njagadish) [inline gbpservice/nfp/configurator/drivers/vpn/vyos/vyos_vpn_driver.py:749]: Let us remove the delete the vpn service delete related calls as they are not needed for GBP case.

File: gbpservice/nfp/configurator/drivers/vpn/vyos/vyos_vpn_driver.py [hunk 7]
```diff
@@ -1100,6 +1101,9 @@ def _get_vm_mgmt_ip_from_desc(self, desc):
         vm_mgmt_ip = tokens[0].split('=')[1]
         return vm_mgmt_ip
 
+    def delete_vpn_service(self, context, resource_data):
```
Comment (njagadish) [inline gbpservice/nfp/configurator/drivers/vpn/vyos/vyos_vpn_driver.py:1104]: Can you run pep8/flake8 tests and unit tests once before commiting?

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py [hunk 8]
```diff
@@ -29,6 +30,9 @@
 
 BASE_VPN_VERSION = '1.0'
 AGENT_TYPE_VPN = 'NFP Vpn agent'
+ACTIVE = 'ACTIVE'
+ERROR = 'ERROR'
+TIMEOUT = 30
```
Comment (njagadish) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py:35]: Make this timeout 20 sec

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py [hunk 9]
```diff
@@ -183,14 +187,46 @@ def create_ipsec_site_connection(self, context, ipsec_site_connection):
...
+                    context,
+                    ipsec_site_connection['vpnservice_id'],
+                    rsrc_type='ipsec_site_connection',
+                    svc_type=self.service_type,
+                    rsrc_id=ipsec_site_connection['id'],
+                    resource=ipsec_site_connection,
+                    reason='create', service_vendor=service_vendor)
+                break
+            elif vpnservice['status'] == ERROR:
+                LOG.error('updating ipsec_site_conn to ERROR state')
```
Comment (njagadish) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py:207]: Can you fix this log. Specify the UUID of ipsec site connection.

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py [hunk 10]
```diff
@@ -183,14 +187,46 @@ def create_ipsec_site_connection(self, context, ipsec_site_connection):
...
+                self._update_ipsec_conn_state(context, ipsec_site_connection)
+                break
+            time.sleep(5)
+            endtime = time.time()
+        else:
+            LOG.error('updating ipsec_site_conn to ERROR state')
+            self._update_ipsec_conn_state(context, ipsec_site_connection)
+
+    def _update_ipsec_conn_state(self, context, ipsec_site_connection):
```
Comment (njagadish) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py:216]: Call it _move_ipsec_conn_state_to_error()

File: gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py (hunk 3 above)
Comment (njagadish) [inline gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaas_driver.py:256]: As discussed, remove this.

**Comment by njagadish** [comment]
Closing this PR as the review comments are addressed and ported to 07 May branch
